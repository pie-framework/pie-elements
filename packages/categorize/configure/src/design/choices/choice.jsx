import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import InputHeader from '../input-header';
import { Checkbox } from '@pie-lib/config-ui';
import { DeleteButton } from '../buttons';
import DragHandle from '@material-ui/icons/DragHandle';
import { DragSource } from 'react-dnd';
import debug from 'debug';
import { uid } from '@pie-lib/drag';

const log = debug('@pie-element:categorize:configure:choice');

const canDrag = props => {
  const count = props.choice.categoryCount || 0;
  if (count === 0) {
    return true;
  } else {
    return props.correctResponseCount < count;
  }
};

export class Choice extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    choice: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    connectDragSource: PropTypes.func.isRequired,
    connectDragPreview: PropTypes.func.isRequired,
    correctResponseCount: PropTypes.number.isRequired,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired
    })
  };

  static defaultProps = {};

  changeContent = content => {
    const { onChange, choice } = this.props;
    choice.content = content;
    onChange(choice);
  };

  changeCategoryCount = () => {
    const { onChange, choice } = this.props;
    if (choice.categoryCount === 1) {
      choice.categoryCount = 0;
    } else {
      choice.categoryCount = 1;
    }
    onChange(choice);
  };

  render() {
    const {
      classes,
      className,
      choice,
      onDelete,
      connectDragSource,
      connectDragPreview,
      imageSupport
    } = this.props;

    const draggable = canDrag(this.props);

    return (
      <Card className={classNames(classes.choice, className)}>
        <CardActions className={classes.actions}>
          {connectDragSource(
            <span
              className={classNames(
                classes.dragHandle,
                draggable === false && classes.dragDisabled
              )}
            >
              <DragHandle color={draggable ? 'primary' : 'disabled'} />
            </span>
          )}
        </CardActions>
        {connectDragPreview(
          <span>
            <InputHeader
              imageSupport={imageSupport}
              label={choice.content}
              onChange={this.changeContent}
              onDelete={onDelete}
            />
          </span>
        )}

        <CardActions className={classes.actions}>
          <DeleteButton label={'delete'} onClick={onDelete} />
          <Checkbox
            mini
            label={'Remove after placing'}
            checked={choice.categoryCount === 1}
            onChange={this.changeCategoryCount}
          />
        </CardActions>
      </Card>
    );
  }
}
const styles = theme => ({
  actions: {
    padding: 0,
    justifyContent: 'space-between'
  },
  choice: {
    padding: theme.spacing.unit,
    overflow: 'visible'
  },
  dragHandle: {
    cursor: 'move'
  },
  dragDisabled: {
    cursor: 'inherit'
  }
});

const StyledChoice = withStyles(styles)(Choice);

export const spec = {
  canDrag,
  beginDrag: props => {
    const out = {
      id: props.choice.id
    };
    log('[beginDrag] out:', out);
    return out;
  },
  endDrag: (props, monitor) => {
    if (!monitor.didDrop()) {
      const item = monitor.getItem();
      if (item.categoryId) {
        log('wasnt droppped - what to do?');
        props.onRemoveChoice(item);
      }
    }
  }
};

const DraggableChoice = DragSource(
  ({ uid }) => uid,
  spec,
  (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    connectDragPreview: connect.dragPreview(),
    isDragging: monitor.isDragging()
  })
)(StyledChoice);

export default uid.withUid(DraggableChoice);
