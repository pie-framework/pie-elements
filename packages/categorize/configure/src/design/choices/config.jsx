import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';

import TextField from '@material-ui/core/TextField';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';
import { Checkbox } from '@pie-lib/config-ui';

const positionOptions = [
  {
    label: 'Above',
    value: 'above',
    secondaryText: 'Choices will be shown above categories'
  },
  {
    label: 'Below',
    value: 'below',
    secondaryText: 'Choices will be shown below categories'
  },
  {
    label: 'Left',
    value: 'left',
    secondaryText: 'Choices will be shown to the left of the categories'
  },
  {
    label: 'Right',
    value: 'right',
    secondaryText: 'Choices will be shown to the right of the categories'
  }
];

export class Config extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    config: PropTypes.object,
    categoryCountIsOne: PropTypes.bool,
    allChoicesHaveCount: PropTypes.func,
    onModelChanged: PropTypes.func
  };

  static defaultProps = {};

  state = {
    anchorEl: null
  };

  toggleRemoveAllTiles = () => {
    const { config, onModelChanged } = this.props;

    const allAtOne = this.props.allChoicesHaveCount(1);
    const update = config.choices.map(c => {
      c.categoryCount = allAtOne ? 0 : 1;
      return c;
    });

    onModelChanged({ choices: update });
  };

  changeLabel = ({ target }) => {
    this.props.onModelChanged({ choicesLabel: target.value });
  };

  toggleShuffle = () => {
    this.props.onModelChanged({
      lockChoiceOrder: !this.props.config.lockChoiceOrder
    });
  };

  changePosition = position => {
    this.props.onModelChanged({ choicesPosition: position.value });

    this.setState({ anchorEl: null });
  };

  handleClickPosition = event => {
    this.setState({ anchorEl: event.currentTarget });
  };

  handleMenuItemClick = (event, index) => {
    this.setState({ selectedIndex: index, anchorEl: null });
  };

  handleClose = () => {
    this.setState({ anchorEl: null });
  };

  render() {
    const { anchorEl } = this.state;
    const { classes, className, categoryCountIsOne, config } = this.props;

    const positionOption =
      positionOptions.find(option => option.value === config.choicesPosition) ||
      positionOption[1];

    return (
      <div className={classNames(classes.config, className)}>
        <div className={classes.row}>
          <Checkbox
            label={'Remove all tiles after placing'}
            checked={categoryCountIsOne}
            onChange={this.toggleRemoveAllTiles}
          />
          <Checkbox
            label={'Lock Choice Order'}
            checked={config.lockChoiceOrder}
            onChange={this.toggleShuffle}
          />
        </div>

        <div className={classes.configuration}>
          <TextField
            className={classes.label}
            InputLabelProps={{
              shrink: true
            }}
            label="Label"
            value={config.choicesLabel}
            onChange={this.changeLabel}
          />
          <List component="nav">
            <ListItem
              button
              aria-haspopup="true"
              onClick={this.handleClickPosition}
            >
              <ListItemText
                primary="Choices Position"
                secondary={positionOption.secondaryText}
              />
            </ListItem>
          </List>
          <Menu
            id="lock-menu"
            anchorEl={anchorEl}
            open={Boolean(anchorEl)}
            onClose={this.handleClose}
          >
            {positionOptions.map(option => (
              <MenuItem
                key={option.value}
                selected={option.value === config.choicesPosition}
                onClick={() => this.changePosition(option)}
              >
                {option.label}
              </MenuItem>
            ))}
          </Menu>
        </div>
      </div>
    );
  }
}
const styles = theme => ({
  configuration: {
    display: 'grid',
    alignItems: 'center',
    gridColumnGap: `${theme.spacing.unit}px`,
    gridRowGap: `${theme.spacing.unit}px`,
    gridTemplateColumns: 'repeat(2, 1fr)',
    justifyContent: 'space-between',
    marginBottom: theme.spacing.unit
  },
  choiceConfig: {
    width: '100%',
    maxWidth: 360,
    backgroundColor: theme.palette.background.paper
  },
  row: {
    display: 'grid',
    gridColumnGap: `${theme.spacing.unit}px`,
    gridRowGap: `${theme.spacing.unit}px`,
    gridTemplateColumns: 'repeat(2, 1fr)',
    justifyContent: 'space-between',
    marginTop: theme.spacing.unit,
    marginBottom: theme.spacing.unit
  },
  shuffleCheckbox: {
    gridColumnStart: 2,
    gridColumnEnd: 2
  }
});
export default withStyles(styles)(Config);
