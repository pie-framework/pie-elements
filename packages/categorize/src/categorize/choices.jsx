import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Choice, { ChoiceType } from './choice';
export { ChoiceType };

const Blank = () => <div />;

export class Choices extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    choices: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.shape(ChoiceType),
        PropTypes.shape({ empty: PropTypes.bool }),
      ])
    ),
    model: PropTypes.shape({
      categoriesPerRow: PropTypes.number,
      choicesLabel: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    choicePosition: PropTypes.string,
  };

  static defaultProps = {
    model: {
      categoriesPerRow: 1,
      choicesLabel: '',
    },
  };

  render() {
    const {
      classes,
      choices = [],
      model,
      disabled,
      choicePosition,
    } = this.props;
    let style = {
      textAlign: 'center',
    };

    if (choicePosition === 'left') {
      style.direction = 'rtl';
    }

    return (
      <div className={classes.wrapper}>
        {model.choicesLabel && model.choicesLabel !== '' && (
          <div className={classes.labelHolder}>{model.choicesLabel}</div>
        )}
        <div className={classes.choices} style={style}>
          {choices.map((c, index) => {
            return c.empty ? (
              <Blank key={index} />
            ) : (
              <Choice
                disabled={disabled}
                className={classes.choice}
                key={index}
                extraStyle={{ maxWidth: `${95 / model.categoriesPerRow}%` }}
                {...c}
              />
            );
          })}
        </div>
      </div>
    );
  }
}

const styles = (theme) => ({
  wrapper: {
    flex: 1,
    padding: theme.spacing.unit,
  },
  choices: {
    padding: theme.spacing.unit / 2,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexWrap: 'wrap'
  },
  labelHolder: {
    margin: '0 auto',
    textAlign: 'center',
    paddingTop: theme.spacing.unit,
  },
});

export default withStyles(styles)(Choices);
