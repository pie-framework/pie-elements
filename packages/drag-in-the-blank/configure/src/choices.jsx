import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import EditableHtml from '@pie-lib/editable-html';
import { renderMath } from '@pie-lib/math-rendering';
import find from 'lodash/find';
import Button from '@material-ui/core/Button';
import Choice from './choice';
import { choiceIsEmpty } from './markupUtils';
import { withStyles } from '@material-ui/core/styles';

window.renMath = renderMath;

const styles = theme => ({
  design: {
    marginTop: theme.spacing.unit * 2
  },
  altChoices: {
    alignItems: 'flex-start',
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-evenly',
    padding: '20px 0 0 0',
    '& > *': {
      marginBottom: '20px'
    }
  }
});

export class Choices extends React.Component {
  static propTypes = {
    duplicates: PropTypes.bool,
    model: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired
  };

  state = {};

  componentDidUpdate() {
    //eslint-disable-next-line
    const domNode = ReactDOM.findDOMNode(this);

    renderMath(domNode);

    if (this.focusedNodeRef) {
      this.focusedNodeRef.focus('end');
    }
  }

  onChoiceChanged = (prevValue, val, key) => {
    const { onChange, model } = this.props;
    const { choices, correctResponse, alternateResponses } = model;
    const newChoices = choices.map(c => {
      if (c.id === key) {
        return { ...c, value: val };
      }

      return c;
    });

    if (choiceIsEmpty({ value: val })) {
      // if the edited content is empty, its usage has to be searched in the correct response definitions
      let usedForResponse = false;

      Object.keys(correctResponse).forEach(responseKey => {
        if (correctResponse[responseKey] === key) {
          usedForResponse = true;
        }
      });

      Object.values(alternateResponses).forEach(alternate => {
        if (alternate.indexOf(key) >= 0) {
          usedForResponse = true;
        }
      });

      if (usedForResponse) {
        alert('Answer choices cannot be blank.');
      } else {
        if (!choiceIsEmpty({ value: prevValue })) {
          // if the previous value was not empty, it means that the choice can be deleted
          const newChoicesWithoutTheEmptyOne = newChoices.filter(choice => choice.id !== key);

          onChange(newChoicesWithoutTheEmptyOne);
        } else {
          onChange(newChoices);
        }
      }
    } else {
      onChange(newChoices);
    }
  };

  onChoiceFocus = id => this.setState({
    focusedEl: id
  });

  onAddChoice = () => {
    const { model: { choices: oldChoices }, onChange } = this.props;

    this.setState({
      focusedEl: `${oldChoices.length}`
    }, () => {
      onChange([
          ...oldChoices,
          {
            id: `${oldChoices.length}`,
            value: ''
          }
        ]
      );
    });
  };

  handleChoiceRemove = id => {
    const { onChange, model: { choices } } = this.props;
    const newChoices = choices.filter(c => c.id !== id);

    onChange(newChoices);
  };

  render() {
    const { focusedEl } = this.state;
    const {
      classes,
      duplicates,
      model: { choices, correctResponse }
    } = this.props;

    const visibleChoices = choices && choices.filter(choice => duplicates || (!duplicates && !find(correctResponse, v => v === choice.id)));

    return (
      <div className={classes.design}>
        <Button
          className={classes.addButton}
          variant="contained"
          color="primary"
          onClick={this.onAddChoice}
        >
          Add Choice
        </Button>
        <div
          className={classes.altChoices}
        >
          {
            visibleChoices &&
            visibleChoices.map((c, index) => {
              if (focusedEl === c.id) {
                return (
                  <div
                    key={index}
                    style={{
                      minWidth: '100%',
                      zIndex: '100'
                    }}
                  >
                    <EditableHtml
                      ref={ref => (this.focusedNodeRef = ref)}
                      className={classes.prompt}
                      markup={c.value}
                      onChange={val => this.onChoiceChanged(c.value, val, c.id)}
                      onDone={() => {
                        this.setState({
                          focusedEl: undefined
                        });
                      }}
                      disableUnderline
                    />
                  </div>
                );
              }

              return (
                <Choice
                  key={index}
                  duplicates={duplicates}
                  targetId="0"
                  choice={c}
                  onClick={() => this.onChoiceFocus(c.id)}
                  onRemoveChoice={() => this.handleChoiceRemove(c.id)}
                />
              );
            })
          }
        </div>
      </div>
    );
  }
}

const Styled = withStyles(styles)(Choices);

export default Styled;
