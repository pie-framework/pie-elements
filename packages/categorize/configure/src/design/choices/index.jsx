import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Divider } from '../buttons';
import classNames from 'classnames';
import Choice from './choice';
import Header from '../header';
import every from 'lodash/every';
import Config from './config';
import { choiceUtils as utils } from '@pie-lib/config-ui';
import { removeAllChoices } from '@pie-lib/categorize';

export class Choices extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    choices: PropTypes.array.isRequired,
    onModelChanged: PropTypes.func.isRequired,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired
    }),
    toolbarOpts: PropTypes.object
  };

  static defaultProps = {};

  changeChoice = choice => {
    const { choices, onModelChanged } = this.props;
    const index = choices.findIndex(h => h.id === choice.id);
    if (index !== -1) {
      choices.splice(index, 1, choice);
      onModelChanged({ choices });
    }
  };

  allChoicesHaveCount = count => {
    const { choices } = this.props;
    return every(choices, c => c.categoryCount === count);
  };

  addChoice = () => {
    const { onModelChanged, model } = this.props;

    const id = utils.firstAvailableIndex(model.choices.map(a => a.id), 0);
    const data = { id, content: 'Choice ' + id };

    onModelChanged({ choices: model.choices.concat([data]) });
  };

  deleteChoice = choice => {
    const { model, onModelChanged } = this.props;
    const index = model.choices.findIndex(a => a.id === choice.id);
    if (index !== -1) {
      model.choices.splice(index, 1);
      model.correctResponse = removeAllChoices(
        choice.id,
        model.correctResponse
      );
      onModelChanged(model);
    }
  };

  render() {
    const {
      classes,
      className,
      choices,
      model,
      imageSupport,
      onModelChanged,
      toolbarOpts
    } = this.props;

    const categoryCountIsOne = this.allChoicesHaveCount(1);
    const choiceHolderStyle = {
      gridTemplateColumns: `repeat(${model.categoriesPerRow}, 1fr)`
    };

    return (
      <div className={classNames(classes.choices, className)}>
        <Header label="Choices" buttonLabel="ADD A CHOICE" onAdd={this.addChoice} />
        <Config
          config={model}
          categoryCountIsOne={categoryCountIsOne}
          onModelChanged={onModelChanged}
          allChoicesHaveCount={this.allChoicesHaveCount}
        />
        <div className={classes.choiceHolder} style={choiceHolderStyle}>
          {choices.map((h, index) => (
            <Choice
              choice={h}
              correctResponseCount={h.correctResponseCount}
              key={index}
              imageSupport={imageSupport}
              onChange={this.changeChoice}
              onDelete={() => this.deleteChoice(h)}
              toolbarOpts={toolbarOpts}
            />
          ))}
        </div>
        <Divider />
      </div>
    );
  }
}
const styles = theme => ({
  choiceHolder: {
    paddingTop: theme.spacing.unit,
    paddingBottom: theme.spacing.unit,
    display: 'grid',
    gridRowGap: `${theme.spacing.unit}px`,
    gridColumnGap: `${theme.spacing.unit}px`
  },
  choices: {
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2
  },
  label: {}
});
export default withStyles(styles)(Choices);
