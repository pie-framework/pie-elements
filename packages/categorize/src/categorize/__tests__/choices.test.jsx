import { shallow } from 'enzyme';
import React from 'react';
import { Choices } from '../choices';

describe('choices', () => {
  const wrapper = (extras) => {
    const defaults = {
      classes: {},
      choices: [],
      onDropChoice: jest.fn(),
      onRemoveChoice: jest.fn(),
      id: '1',
      label: 'Category Label',
      grid: { columns: 1, rows: 1 },
    };

    const props = { ...defaults, ...extras };
    return shallow(<Choices {...props} />);
  };

  describe('snapshots', () => {
    it('renders', () => {
      expect(wrapper()).toMatchSnapshot();
    });

    it('disabled', () => {
      expect(wrapper({ disabled: true })).toMatchSnapshot();
    });

    it('empty', () => {
      expect(wrapper({ choices: [{ id: '1', label: 'foo' }] })).toMatchSnapshot();
    });

    it('empty', () => {
      expect(wrapper({ choices: [{ empty: true }] })).toMatchSnapshot();
    });
  });
});
