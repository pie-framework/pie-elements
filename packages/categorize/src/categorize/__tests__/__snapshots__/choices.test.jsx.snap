// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choices snapshots disabled 1`] = `
<div>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  />
</div>
`;

exports[`choices snapshots empty 1`] = `
<div>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <DragSource(Choice)
      extraStyle={
        Object {
          "maxWidth": "95%",
        }
      }
      id="1"
      key="0"
      label="foo"
    />
  </div>
</div>
`;

exports[`choices snapshots empty 2`] = `
<div>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <Blank
      key="0"
    />
  </div>
</div>
`;

exports[`choices snapshots renders 1`] = `
<div>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  />
</div>
`;
